### PR TITLE
fix(ui): Clear environments when changing projects [SEN-661]

### PR DESCRIPTION
**Pull request by billyvg**
If you are on project A that has envs 1, 2, 3, and you select env 3. If you switch to a project B, there is no guarantee project B will have the same envs 1, 2, 3.

We should clear environments when switching projects since environments are scoped to a project.

Fixes SEN-661

### CODE REVIEW

File: src/sentry/static/sentry/app/components/organizations/globalSelectionHeader/index.jsx [hunk 1]
```diff
@@ -301,7 +301,8 @@ class GlobalSelectionHeader extends React.Component {
   handleUpdateProjects = () => {
     const {projects} = this.state;
     updateProjects(projects, this.getRouter(), this.getUpdateOptions());
-    this.setState({projects: null});
+    updateEnvironments(null, this.getRouter(), this.getUpdateOptions());
+    this.setState({projects: null, environments: null});
```
Comment (markstory) [inline src/sentry/static/sentry/app/components/organizations/globalSelectionHeader/index.jsx]: Didn't we have code that looked for intersections between the currently selected environments and the environments belonging to the newly selected projects?

Comment (billyvg) [reply src/sentry/static/sentry/app/components/organizations/globalSelectionHeader/index.jsx]: Ah we do, it just doesn't completely work

File: src/sentry/static/sentry/app/components/organizations/multipleEnvironmentSelector.jsx [hunk 2]
```diff
@@ -80,6 +80,7 @@ class MultipleEnvironmentSelector extends React.PureComponent {
   replaceSelected(newSelection) {
     this.setState({selectedEnvs: new Set(newSelection)});
     this.props.onChange(newSelection);
+    this.doUpdate();
```
Comment (billyvg) [inline src/sentry/static/sentry/app/components/organizations/multipleEnvironmentSelector.jsx]: @markstory Internal state of the dropdown was being updated, but change was never saved/propagated (the environment would still exist in URL + store)

Comment (markstory) [reply src/sentry/static/sentry/app/components/organizations/multipleEnvironmentSelector.jsx]: I found https://github.com/getsentry/sentry/commit/1aaab50d7a4fb95023a06cc003daf99ce78e1a36#diff-0c4ce644581f429b0c68084cf1f86a35 in the commit logs, and the parts about `list -> details` navigation having issues might be something to check again.

Comment (billyvg) [reply src/sentry/static/sentry/app/components/organizations/multipleEnvironmentSelector.jsx]: Hmm, is it better to go back to the original logic where we clear environments on an explicit project change? It seems unlikely you'd want to keep an environment selected when you change your project(s) context

Comment (markstory) [reply src/sentry/static/sentry/app/components/organizations/multipleEnvironmentSelector.jsx]: Given the numerous sharp edges it might be simplest.

**Comment by logicalbomb** [comment]
Where did we end up with this conversation?

File: src/sentry/static/sentry/app/components/organizations/multipleEnvironmentSelector.jsx (hunk 2 above)
Comment (billyvg) [reply src/sentry/static/sentry/app/components/organizations/multipleEnvironmentSelector.jsx]: @markstory 	Updated to clear envs when we change projects https://github.com/getsentry/sentry/pull/13334/commits/15ea6fd3a0591a19d05d74a873170f3faca07c4d

**Comment by logicalbomb** [comment]
@billyvg @markstory Let's get this cleaned up and shipped soon. You have product approval.

File: src/sentry/static/sentry/app/components/organizations/globalSelectionHeader/index.jsx (hunk 1 above)
Comment (markstory) [reply src/sentry/static/sentry/app/components/organizations/globalSelectionHeader/index.jsx]: Ok. Should we remove it? It lives https://github.com/getsentry/sentry/blob/26939bf07cae41820d8d793e9a588b09f6bc117e/src/sentry/static/sentry/app/components/organizations/multipleEnvironmentSelector.jsx#L62-L78

File: src/sentry/static/sentry/app/components/organizations/multipleEnvironmentSelector.jsx [hunk 3]
```diff
@@ -60,27 +60,14 @@ class MultipleEnvironmentSelector extends React.PureComponent {
...
-      const selectedEnvs = Array.from(this.state.selectedEnvs.values());
-      const validEnvironments = this.getEnvironments();
-      const newSelection = validEnvironments.filter(env => selectedEnvs.includes(env));
-
-      if (newSelection.length !== selectedEnvs.length) {
-        this.replaceSelected(newSelection);
-      }
+    // Need to sync state
+    if (this.props.value !== prevProps.value) {
+      this.syncSelectedStateFriomProps();
```
Comment (markstory) [inline src/sentry/static/sentry/app/components/organizations/multipleEnvironmentSelector.jsx]:
```suggestion
      this.syncSelectedStateFromProps();
```

File: src/sentry/static/sentry/app/components/organizations/multipleEnvironmentSelector.jsx [hunk 4]
```diff
@@ -60,27 +60,14 @@ class MultipleEnvironmentSelector extends React.PureComponent {
...
     }
   }
 
-  replaceSelected(newSelection) {
-    this.setState({selectedEnvs: new Set(newSelection)});
-    this.props.onChange(newSelection);
-  }
+  syncSelectedStateFriomProps = () =>
```
Comment (markstory) [inline src/sentry/static/sentry/app/components/organizations/multipleEnvironmentSelector.jsx]:
```suggestion
  syncSelectedStateFromProps = () =>
```

File: tests/js/spec/components/organizations/multipleEnvironmentSelector.spec.jsx [hunk 5]
```diff
@@ -58,32 +58,15 @@ describe('MultipleEnvironmentSelector', function() {
...
-    wrapper.setProps({selectedProjects: [2]});
-    await wrapper.update();
-
-    expect(onChange).toHaveBeenCalled();
-    const selector = wrapper.find('MultipleEnvironmentSelector').instance();
-    expect(selector.state.selectedEnvs).toEqual(new Set([]));
+    // wrapper
+    // .find('MultipleEnvironmentSelector')
+    // .instance()
+    // .doUpdate();
```
Comment (markstory) [inline tests/js/spec/components/organizations/multipleEnvironmentSelector.spec.jsx]: Some commented out code here.

File: tests/js/spec/components/organizations/multipleEnvironmentSelector.spec.jsx [hunk 6]
```diff
@@ -58,32 +58,15 @@ describe('MultipleEnvironmentSelector', function() {
     expect(onChange).toHaveBeenCalledTimes(3);
     expect(onChange).toHaveBeenLastCalledWith(envs, expect.anything());
 
-    wrapper.setProps({value: envs});
-    wrapper.update();
+    // console.log(wrapper.find('MultipleSelectorSubmitRow').debug());
```
Comment (markstory) [inline tests/js/spec/components/organizations/multipleEnvironmentSelector.spec.jsx]: Some commented out code here.

**Review by markstory — APPROVED** [review]

**Pull request merged**